File: sdk/c#/PicarioXPO.RenderAPI.V2/TextKeys.cs
using System.Collections.Generic;
using System.Linq;
using System.Text;
using PicarioXPO.RenderAPI.V2.TextKeyExtractors;

namespace PicarioXPO.RenderAPI.V2
{
    internal class TextKeysCollection
    {
        internal List<TextKey> Keys
        {
            get
            {
                return keys;
            }
        }

        private readonly List<TextKey> keys = new List<TextKey>
        {
            new TextTextKey(),
            new TextFontTextKey(),
            new TextColorTextKey(),
            new TextSizeTextKey(),
            new TextAlignmentTextKey(),
            new TextStyleTextKey(),
            new TextPlacingPointXTextKey(),
            new TextPlacingPointYTextKey(),
            new TextMultiplierTextKey(),
            new TextRotationTextKey()
        };
    }


    internal static class TextKeys
    {
        public static StringBuilder AppendTexts(this StringBuilder stringBuilder, IEnumerable<XpoUrlObject> xpoUrlTexts)
        {
            var xpoUrlObjects = xpoUrlTexts as XpoUrlObject[] ?? xpoUrlTexts.ToArray();

            if (!xpoUrlObjects.Any())
                return stringBuilder;

            var textKeys = new TextKeysCollection();

            foreach (var textKey in textKeys.Keys)
            {
                var keyValue = textKey.GetValues(xpoUrlObjects);

                if (!string.IsNullOrEmpty(keyValue))
                {
                    stringBuilder.Append("&");
                    stringBuilder.Append(keyValue);
                }
            }

            return stringBuilder;
        }
    }
}

File: sdk/js/PicarioXPO.RenderAPI.V2/XpoUrlGenerator/TextKeys.ts
class TextKeys {
    private textKeys = new Array<TextKey>();

    constructor() {
        this.textKeys.push(new TextTextKey(),
                           new TextColorTextKey(),
                           new TextAlignmentTextKey(),
                           new TextFontTextKey(),
                           new TextMultiplierTextKey(),
                           new TextPlacingPointXTextKey(),
                           new TextPlacingPointYTextKey(),
                           new TextSizeTextKey(),
                           new TextStyleTextKey());
    }


    appendTexts(stringBuilder: string, xpoUrlTexts: Array<XpoUrlObject>) {
        if (xpoUrlTexts == null || xpoUrlTexts.length <= 0)
            return stringBuilder;

        for (var i = 0; i < this.textKeys.length; i++) {
            var keyValue = this.textKeys[i].getValues(xpoUrlTexts);

            if (keyValue) {
                stringBuilder = stringBuilder.concat("&")
                                             .concat(keyValue);
            }
        }

        return stringBuilder;
    }
    
} 